extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/impl/DeviceImpl.java

```diff
@@ -0,0 +1,1859 @@
+                        new Short[] { sceneNumber, prio }));
+            }
+        }
+    }
+
+    @Override
+    public synchronized void undoInternalScene(InternalScene scene) {
+        logger.debug("undo Scene {} dSID {}", scene.getSceneID(), dsid.getValue());
+        if (activeScene != null && activeScene.equals(scene)) {
+            if (lastCallWasUndo) {
+                lastScene = null;
+                return;
+            }
+            if (this.lastScene != null && !lastScene.equals(activeScene)) {
+                activeScene = lastScene;
+                lastScene = null;
+                activeScene.activateSceneByDevice();
+            } else {
+                internalUndoScene();
+                logger.debug("internalUndo Scene dSID {}", dsid.getValue());
+                this.activeScene = null;
+            }
+            internalUndoScene();
+            lastCallWasUndo = true;
+        }
+    }
+
+    @Override
+    public synchronized void callScene(Short sceneNumber) {
+        this.deviceStateUpdates.add(new DeviceStateUpdateImpl(DeviceStateUpdate.UPDATE_CALL_SCENE, sceneNumber));
+    }
+
+    @Override
+    public synchronized void internalCallScene(Short sceneNumber) {
+        logger.debug("call Scene id {} dSID {}", sceneNumber, dsid.getValue());
+        if (isDeviceWithOutput()) {
+            activeSceneNumber = sceneNumber;
+            informLastSceneAboutSceneCall(sceneNumber);
+            if (!isShade()) {
+                outputValueBeforeSceneCall = this.outputValue;
+            } else {
+                outputValueBeforeSceneCall = this.slatPosition;
+                if (isBlind()) {
+                    slatAngleBeforeSceneCall = this.slatAngle;
+                }
+            }
+            if (!checkSceneNumber(sceneNumber)) {
+                if (containsSceneConfig(sceneNumber)) {
+                    if (doIgnoreScene(sceneNumber)) {
+                        return;
+                    }
+                } else {
+                    this.deviceStateUpdates.add(new DeviceStateUpdateImpl(DeviceStateUpdate.UPDATE_SCENE_CONFIG,
+                            new Short[] { sceneNumber, 0 }));
+                }
+                if (sceneOutputMap.get(sceneNumber) != null) {
+                    if (!isShade()) {
+                        updateInternalDeviceState(new DeviceStateUpdateImpl(DeviceStateUpdate.OUTPUT,
+                                sceneOutputMap.get(sceneNumber)[0]));
+                    } else {
+                        updateInternalDeviceState(new DeviceStateUpdateImpl(DeviceStateUpdate.SLATPOSITION,
+                                sceneOutputMap.get(sceneNumber)[0]));
+                        if (isBlind()) {
+                            updateInternalDeviceState(new DeviceStateUpdateImpl(DeviceStateUpdate.SLAT_ANGLE,
+                                    sceneOutputMap.get(sceneNumber)[1]));
```

Comment:
this looks like another "magic contract", who defines that position 1 of `sceneOutputMap.get(sceneNumber)` is responsible for blinds? A named reference would be more readable and maintainable...